docs/AUTH_PROTOCOL.md

```diff
@@ -0,0 +1,393 @@
+
+Once gnome-shell recieves the FD, all it has to do is add it to the main loop (i.e. `poll` it) and start reading from the FD to recieve events.
+
+Now, for the duration of the conversation PAM will be blocking and waiting for a response. The request that sent over the FD was, in PAM's view, a question. So the answer must come from the frontend before anything else can happen. Of course the authentication is happening out of band. Once the frontend decides that it's done with authentication (i.e. if it has a fatal `authenticationFailed` reponse, or `authenticationSuccessful`), then it can send an empty PAM "answer" to the "question". PAM then takes back over, and either returns a PAM error or continues and allows the gdm-session-worker to start a session.
+
+
+
+## Startup
+
+The protocol begins with a set sequence of events, which follows. First the authenticator sends:
+
+```js
+{
+    "event": "hello",
+    "version": 1,
+}
+```
+
+If the version is unsupported, the frontend should just close the socket and then send PAM an empty answer so that PAM can take over. Since there hasn't yet been an `authenticationSuccessful` event, PAM will return a failure.
+
+If the version is supported, the frontend reponds:
+
+```js
+{
+    "event": "hello",
+    "supportedMechanisms": [ ... ],
+    "supportedAuthFailureReasons": [ ... ],
+}
+```
+
+This communicates the frontend's capabilities to the authenticator. Mechanisms and authentication failure reasons are documented below.
+
+In response to this information, the authenticator responds with a series of "flows". A flow is a sequence of mechanisms. Most commonly, a flow would correspond 1:1 to some mechanism (i.e. the `password` mechanism). However, flows can be more complicated: for instance, your enterprise might want to define a flow that first prompts for a `password`, then does `eidp` SSO login, and finally prompts for an `otp`. 
+
+Flows exist to give the user a choice of how they want to log in. The frontend will present a list of flows to the user, and the user can select whichever flow they want to attempt authentication with. Most commonly, this will be basic stuff: does the user want to log in via "Password" or "Pin" or "Smartcard". But again, flows are flexible enough that they can represent more complicated sequences.
+
+Flows are communicated as follows:
+
+```js
+{
+    "event": "flows",
+    "flows": [
+        { // This is a single flow
+            "id": "mycorp-login-flow",
+            "primaryMechanism": "eidp",
+            "name": "MyCorp Login",
+        },
+        ... // There can be more flows that follow
+    ]
+}
+```
+
+The authenticator gives the frontend an ordered list of flows. The flows are ordered in "most-recommended" order (as chosen by the authenticator), so if the frontend needs to pick a flow to start by default it should pick the first flow in the list.
```

Comment:
> so if the frontend needs to pick a flow to start by default it should pick the first flow in the list

Since the frontend can remember which flow the user prefers (as stated in the next paragraph), shouldn't it be able to choose the user-preferred flow as the default?

Maybe make it clear that the frontend should only pick the first flow if there is no user-preferred flow if that's how it should work.